jujugui/static/gui/src/app/state/state.js

```diff
@@ -117,30 +149,192 @@ const State = class State {
+
+  /**
     Checks the current location and parses it, building the state, then
     executing the registered dispatchers.
+    @param {Array} nullKeys - A list of keys which we must run the 'cleanup'
+      dispaches on before dispatching the current state.
+    @param {Boolean} updateState - Whether this dispatch should update the
+      internal app state. This is typically only done when dispatch is called
+      manually. Internal calls to dispatch are done via changeState which
+      updates the state itself so it's not needed to do it here.
   */
-  dispatch() {
+  dispatch(nullKeys = [], updateState = true) {
     let error, state;
     ({error, state} = this.generateState(this.location.href));
-    this._appStateHistory.push(state);
     if (error !== null) {
-      return `unable to generate state: ${error}`;
+      error += `unable to generate state: ${error}`;
+      return {error, state};
+    }
+
+    /**
+      Extract out the state path for the dispatcher key.
+      @param {Object} state - The app state.
+    */
+    function extract(state) {
+      let allKeys = [];
+      function concat(state, keys = []) {
+        Object.keys(state).forEach(key => {
+          keys.push(key);
+          if (typeof state[key] === 'object' && state[key] !== null) {
+            concat(state[key], keys);
+          } else {
+            allKeys.push(keys.join('.'));
+          }
+          keys = [];
+        });
+      }
+      concat(state);
+      return allKeys;
     }
-    this._dispatch(state);
+    if (updateState) {
+      this._appStateHistory.push(state);
+    }
+    // First run all of the 'null state' dispatchers to clear out the old
+    // state representations.
+    nullKeys.forEach(key => this._dispatch(state, key, true));
+    // Then execute the 'all' dispatchers.
+    this._dispatch(state, '*');
+    // Extract and loop through the state keys to execute their dispatchers.
+    extract(state).forEach(key => {
+      this._dispatch(state, key);
+    });
+
+    return {error: null, state};
   }
 
   /**
     Takes the existing app state and then calls the registered dispatchers.
+    @param {Object} state - The state to dispatch.
+    @param {String} key - The key to manage the dispatchers.
+    @param {Boolean} cleanup - Whether it should execute the cleanup method or
+      not. Defaults to false.
   */
-  _dispatch() {}
+  _dispatch(state, key, cleanup = false) {
+    if (!this._dispatchers[key]) {
+      // If we have no registered dispatchers for the key then return.
+      return;
+    }
+    const iterator = this._dispatchers[key][Symbol.iterator]();
+    function next() {
+      const data = iterator.next();
+      if (!data.done) {
+        const index = cleanup ? 1 : 0;
+        // The 0 index is the 'create' dispatcher, the 1 index is the
+        // 'cleanup' dispatcher.
+        const dispatcher = data.value[index];
+        if (typeof dispatcher === 'function') {
+          dispatcher(state, next);
+        }
```

Comment:
I didn't want to force the user to create dispatchers for cleaning up so this is just a simple way of ignoring it if one wasn't provided.